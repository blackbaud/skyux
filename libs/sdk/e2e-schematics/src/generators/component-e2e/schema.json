{
  "$schema": "http://json-schema.org/schema",
  "$id": "library-e2e",
  "title": "Create a library e2e",
  "description": "Creates an e2e infrastructure for a library.",
  "type": "object",
  "properties": {
    "name": {
      "type": "string",
      "description": "The name of the library.",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What is the name of the library?",
      "pattern": "^[a-zA-Z].*$"
    },
    "tags": {
      "type": "string",
      "description": "Add tags to the library (used for linting)."
    }
  },
  "additionalProperties": false,
  "required": ["name"]
}
