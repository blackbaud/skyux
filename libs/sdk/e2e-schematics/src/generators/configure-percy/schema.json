{
  "$schema": "http://json-schema.org/schema",
  "$id": "configure-percy",
  "type": "object",
  "properties": {
    "name": {
      "type": "string",
      "description": "e2e project name",
      "$default": {
        "$source": "argv",
        "index": 0
      }
    }
  }
}
