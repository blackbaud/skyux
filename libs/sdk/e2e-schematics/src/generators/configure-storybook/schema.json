{
  "$schema": "http://json-schema.org/schema",
  "$id": "configure-storybook",
  "title": "Configure Storybook",
  "description": "Adds configuration to Storybook.",
  "type": "object",
  "properties": {
    "name": {
      "type": "string",
      "description": "The name of the project using Storybook.",
      "$default": {
        "$source": "argv",
        "index": 0
      }
    }
  },
  "additionalProperties": false
}
