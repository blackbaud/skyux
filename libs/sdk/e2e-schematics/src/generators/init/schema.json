{
  "$schema": "http://json-schema.org/schema",
  "cli": "nx",
  "$id": "Init",
  "title": "Configure E2E environment"
}
