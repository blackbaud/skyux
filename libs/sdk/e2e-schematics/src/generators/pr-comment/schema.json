{
  "$schema": "http://json-schema.org/schema",
  "$id": "PrComment",
  "title": "PR Comment",
  "description": "Generate markdown for a PR comment or a PR preview README.md",
  "type": "object",
  "properties": {
    "pr": {
      "type": "number",
      "description": "PR Number",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "Which PR is this for?"
    },
    "storybooks": {
      "type": "string",
      "description": "Comma separated list of storybooks to include in the comment",
      "alias": "s",
      "x-prompt": "Which storybook projects should be listed?"
    },
    "apps": {
      "type": "string",
      "description": "Comma separated list of apps to include in the comment",
      "alias": "a"
    }
  },
  "required": ["pr", "storybooks"]
}
