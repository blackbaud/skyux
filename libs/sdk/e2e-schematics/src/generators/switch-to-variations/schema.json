{
  "$schema": "http://json-schema.org/schema",
  "$id": "SwitchToVariations",
  "title": "",
  "type": "object",
  "properties": {
    "project": {
      "type": "string",
      "description": "E2E Project"
    }
  },
  "required": ["project"]
}
