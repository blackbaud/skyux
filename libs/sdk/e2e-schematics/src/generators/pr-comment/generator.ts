import { Tree, generateFiles, joinPathFragments } from '@nx/devkit';

import { PrCommentGeneratorOptions, PrCommentGeneratorSchema } from './schema';

function normalizeOptions(
  tree: Tree,
  options: PrCommentGeneratorSchema
): PrCommentGeneratorOptions {
  const pr = options.pr;
  const storybooks = options.storybooks
    .split(',')
    .map((s) => s.trim())
    .filter((s) => s.length > 0);
  const apps = (options.apps || '')
    .split(',')
    .map((s) => s.trim())
    .filter((s) => s.length > 0);
  if (!pr || isNaN(parseInt(`${pr}`, 10))) {
    throw new Error('Invalid PR number');
  }
  return {
    pr,
    storybooks,
    apps,
  };
}

export default async function prComment(
  tree: Tree,
  options: PrCommentGeneratorSchema
) {
  const normalizedOptions = normalizeOptions(tree, options);
  generateFiles(
    tree,
    joinPathFragments(__dirname, 'files/comment'),
    'dist',
    normalizedOptions
  );
  generateFiles(
    tree,
    joinPathFragments(__dirname, 'files/readme'),
    'dist',
    normalizedOptions
  );
}
