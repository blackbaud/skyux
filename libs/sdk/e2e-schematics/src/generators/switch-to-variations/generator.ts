import { Tree, getProjects, logger, visitNotIgnoredFiles } from '@nx/devkit';

import { SwitchToVariationsGeneratorSchema } from './schema';

export default async function (
  tree: Tree,
  options: SwitchToVariationsGeneratorSchema
) {
  const projectConfiguration = getProjects(tree).get(options.project);
  if (!projectConfiguration) {
    throw new Error(`Project ${options.project} not found`);
  }

  // istanbul ignore if
  if (!(projectConfiguration.sourceRoot && projectConfiguration.targets)) {
    throw new Error(`Project source root or targets could not be found`);
  }

  if (!projectConfiguration.targets['e2e']) {
    throw new Error(`Project ${options.project} is not an e2e project`);
  }

  visitNotIgnoredFiles(tree, projectConfiguration.sourceRoot, (filepath) => {
    let reason = '';
    if (filepath.endsWith('.spec.ts') || filepath.endsWith('.cy.ts')) {
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      const source = tree.read(filepath, 'utf-8')!;
      // Quick check if this looks like the old pattern
      if (
        source.startsWith(
          `['default', 'modern-light', 'modern-dark'].forEach((theme) => {`
        )
      ) {
        const describePattern =
          /^ {2}describe\(`([-\w ]+) in \$\{theme} theme`, \(\) => \{$/;
        const specLines = source.split(/\r?\n/);
        specLines.shift(); // Remove first line
        const describeLine = `${specLines.shift()}`;
        if (describeLine.match(describePattern)) {
          // Strip empty lines from the bottom of the file
          while (specLines[specLines.length - 1].trim() === '') {
            specLines.pop();
          }
          const lastLine = specLines.pop();
          // Is this a single block?
          if (
            lastLine === '});' &&
            specLines.every((line) => line.startsWith('  ') || line === '')
          ) {
            const newSpecLines = [
              `import { E2eVariations } from '@skyux-sdk/e2e-schematics';`,
              '',
              describeLine.replace(
                describePattern,
                (_, description) => `describe(\`${description}\`, () => {`
              ),
              '  E2eVariations.forEachTheme((theme) => {',
              '    describe(`in ${theme} theme`, () => {',
            ];
            newSpecLines.push(
              ...specLines.map((line) => (line === '' ? '' : `  ${line}`)),
              '  });',
              '});',
              ''
            );
            tree.write(filepath, newSpecLines.join(`\n`));
            return;
          } else if (lastLine === '});') {
            reason = 'multiple blocks';
          } else {
            reason = `last line is ${lastLine}`;
          }
        } else {
          reason = `describe line does not match: ${describeLine}`;
        }
        logger.info(`Unable to update ${filepath}: ${reason}`);
      }
    }
  });
}
