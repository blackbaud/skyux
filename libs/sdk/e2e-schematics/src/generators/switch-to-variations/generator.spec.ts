import { applicationGenerator } from '@nx/angular/generators';
import {
  NxJsonConfiguration,
  Tree,
  generateFiles,
  joinPathFragments,
  readNxJson,
  updateNxJson,
} from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';

import generator from './generator';
import { SwitchToVariationsGeneratorSchema } from './schema';

describe('switch-to-variations generator', () => {
  let appTree: Tree;
  const options: SwitchToVariationsGeneratorSchema = { project: 'test-e2e' };

  beforeEach(async () => {
    appTree = createTreeWithEmptyWorkspace({ layout: 'apps-libs' });
    const nxJson: NxJsonConfiguration = readNxJson(appTree) || {};
    nxJson.workspaceLayout = {
      appsDir: 'apps',
      libsDir: 'libs',
    };
    updateNxJson(appTree, nxJson);
    appTree.write('.gitignore', '# test');
    await applicationGenerator(appTree, {
      name: 'test',
    });
  });

  it('should run successfully', async () => {
    const start = appTree.read(
      'apps/test-e2e/src/integration/app.spec.ts',
      'utf-8'
    );
    const changes = appTree.listChanges();
    await generator(appTree, options);
    expect(
      appTree.read('apps/test-e2e/src/integration/app.spec.ts', 'utf-8')
    ).toEqual(start);
    expect(appTree.listChanges()).toEqual(changes);
  });

  it('should throw errors', async () => {
    try {
      await generator(appTree, {
        ...options,
        project: 'bogus',
      });
      fail('should have thrown');
    } catch (e) {
      if (!(e instanceof Error)) {
        fail('should have thrown an Error');
      }
      expect(e.message).toContain('Project bogus not found');
    }
    try {
      await generator(appTree, {
        ...options,
        project: 'test',
      });
      fail('should have thrown');
    } catch (e) {
      if (!(e instanceof Error)) {
        fail('should have thrown an Error');
      }
      expect(e.message).toContain('Project test is not an e2e project');
    }
  });

  it('should update test', async () => {
    generateFiles(
      appTree,
      joinPathFragments(__dirname, 'files/fixtures'),
      'apps/test-e2e/src/integration',
      {}
    );
    await generator(appTree, options);
    expect(
      appTree.read('apps/test-e2e/src/integration/example1.spec.ts', 'utf-8')
    ).toMatchSnapshot();
  });
});
