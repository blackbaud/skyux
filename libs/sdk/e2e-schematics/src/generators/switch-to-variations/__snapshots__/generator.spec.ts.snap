// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`switch-to-variations generator should update test 1`] = `
"import { E2eVariations } from '@skyux-sdk/e2e-schematics';

describe(\`ag-grid-storybook data manager\`, () => {
  E2eVariations.forEachTheme((theme) => {
    describe(\`in \${theme} theme\`, () => {
      [
        ['normal', 'normal'],
        ['normal-with-top-scroll', 'normal with top scroll'],
        ['auto-height', 'auto height'],
        ['auto-height-with-top-scroll', 'auto height with top scroll'],
      ].forEach(([domLayout, label]) => {
        describe(\`\${label} layout\`, () => {
          beforeEach(() => {
            cy.viewport(1300, 900).visit(
              \`/iframe.html?globals=theme:\${theme}&id=datamanagercomponent-datamanager--data-manager-\${domLayout}\`
            );
            cy.waitForFaAndBbFonts();
          });

          it(\`should render ag-grid with data manager, \${label} layout\`, () => {
            cy.get('#ready')
              .should('exist')
              .end()
              .get('#storybook-root')
              .should('exist')
              .should('be.visible')
              .screenshot(
                \`datamanagercomponent-datamanager--data-manager-\${domLayout}-\${theme}\`
              )
              .percySnapshot(
                \`datamanagercomponent-datamanager--data-manager-\${domLayout}-\${theme}\`,
                {
                  widths: [1280],
                  scope: '#storybook-root',
                  percyCSS: \`
                    /* Avoid \\"virtual rows\\" in the screenshot. */
                    :root {
                      --viewport-height: 600px;
                    }
                  \`,
                }
              );
          });
        });
      });
    });
  });
});
"
`;
